src/ORM/RelatedData/StandardRelatedDataService.php

```diff
@@ -0,0 +1,507 @@
+
+    /**
+     * @var DataObjectSchema
+     */
+    private $dataObjectSchema;
+
+    /**
+     * @var array
+     */
+    private $classToTableName;
+
+    /**
+     * Find all DataObject instances that have a linked relationship with $record
+     *
+     * @param DataObject $record
+     * @param string[] $excludedClasses
+     * @return SS_List
+     */
+    public function findAll(DataObject $record, array $excludedClasses = []): SS_List
+    {
+        // Do not query unsaved DataObjects
+        if (!$record->exists()) {
+            return ArrayList::create();
+        }
+
+        $this->config = Config::inst()->getAll();
+        $this->dataObjectSchema = DataObjectSchema::create();
+        $this->initClassToTableName();
+        $classIDs = [];
+        $throughClasses = [];
+
+        // "regular" relations i.e. point from $record to different DataObject
+        $this->addRelatedHasOnes($classIDs, $record);
+        $this->addRelatedManyManys($classIDs, $record, $throughClasses);
+
+        // Loop config data to find "reverse" relationships pointing back to $record
+        foreach (array_keys($this->config) as $lowercaseClassName) {
+            if (!class_exists($lowercaseClassName)) {
+                continue;
+            }
+            // Example of $class: My\App\MyPage (extends SiteTree)
+            try {
+                $class = ClassInfo::class_name($lowercaseClassName);
+            } catch (ReflectionException $e) {
+                continue;
+            }
+            if (!is_subclass_of($class, DataObject::class)) {
+                continue;
+            }
+            $this->addRelatedReverseHasOnes($classIDs, $record, $class);
+            $this->addRelatedReverseManyManys($classIDs, $record, $class, $throughClasses);
+        }
+        $this->removeClasses($classIDs, $excludedClasses, $throughClasses);
+        $classObjs = $this->fetchClassObjs($classIDs);
+        return $this->deriveList($classIDs, $classObjs);
+    }
+
+    /**
+     * Loop has_one relationships on the DataObject we're getting usage for
+     * e.g. File.has_one = Page, Page.has_many = File
+     *
+     * @param array $classIDs
+     * @param DataObject $record
+     */
+    private function addRelatedHasOnes(array &$classIDs, DataObject $record): void
+    {
+        $class = get_class($record);
+        foreach ($record->hasOne() as $component => $componentClass) {
+            $componentIDField = "{$component}ID";
+            $tableName = $this->findTableNameContainingComponentIDField($class, $componentIDField);
+            if ($tableName === '') {
+                continue;
+            }
+
+            $select = sprintf('"%s"', $componentIDField);
+            $where = sprintf('"ID" = %u AND "%s" > 0', $record->ID, $componentIDField);
+
+            // Polymorphic
+            // $record->ParentClass will return null if the column doesn't exist
+            if ($componentIDField === 'ParentID' && $record->ParentClass) {
```

Comment:
I'm a bit late to this party, but if you want to detect polymorphic has_ones, you should be checking `$componentClass === DataObject::class`.

This code has a incorrect baked-in assumption that ONLY relations named ParentID can be polymorphic, when any relation can be polymorphic at all. E.g. `RelatedID, RelatedClass`.

The code could even have a `getParentClass()` method that would trick this code into thinking there's a polymorphic relation too.